### PR TITLE
dnn : Fix convolution asymmetric padding bug in onnx importer

### PR DESCRIPTION
**Pull request by SamFC10**
**Merge with extra** : opencv/opencv_extra#881
 
relates #19316 #19499 #19696 #19957

Using tools like tf2onnx and keras2onnx, to convert tensorflow networks to onnx format, fuses zero-padding layer with convolution layer but does not check if the padding is symmetric when importing such converted networks.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2021.4.0:20.04
build_image:Custom Win=openvino-2021.4.0
build_image:Custom Mac=openvino-2021.4.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*Conv*
```

### CODE REVIEW

File: modules/dnn/test/test_onnx_importer.cpp [hunk 1]
```diff
@@ -109,6 +109,7 @@ TEST_P(Test_ONNX_layers, MaxPooling_2)
 TEST_P(Test_ONNX_layers, Convolution)
 {
     testONNXModels("convolution");
+    testONNXModels("conv_asymmetric_pads");
```
Comment (alalek) [inline modules/dnn/test/test_onnx_importer.cpp:112]: Is it possible to provide ONNX test data generator code ([`generate_onnx_models.py`](https://github.com/opencv/opencv_extra/blob/master/testdata/dnn/onnx/generate_onnx_models.py) in opencv_extra PR) ?

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/dnn/test/test_onnx_importer.cpp (hunk 1 above)
Comment (SamFC10) [reply modules/dnn/test/test_onnx_importer.cpp:112]: No, I don't think its possible. torch.onnx.export() does not fuses padding with conv layer which is required to test this change.
As mentioned in description, tools like tf2onnx and keras2onnx use this type of fusion (onnx/tensorflow-onnx#334). The onnx model in opencv_extra has been generated using tf2onnx.

Comment (rogday) [reply modules/dnn/test/test_onnx_importer.cpp:112]: > The onnx model in opencv_extra has been generated using tf2onnx.

tf2onnx has python interface, would it be possible to use that?

Comment (SamFC10) [reply modules/dnn/test/test_onnx_importer.cpp:112]: > tf2onnx has python interface

Yes it can be used, but the original network is built using tensorflow. `generate_onnx_models.py` uses Pytorch.
Code used to generate onnx model
```python3
import tensorflow as tf
import tf2onnx
tf.keras.backend.set_image_data_format('channels_first')

inp = tf.keras.layers.Input(batch_shape = (1, 2, 3, 4))
pad = tf.keras.layers.ZeroPadding2D(padding = ((2, 1), (2, 1)))(inp)
out = tf.keras.layers.Conv2D(3, kernel_size = 3)(pad)
model = tf.keras.Model(inp, out)
onnx_model, _ = tf2onnx.convert.from_keras(model, output_path="/conv_asymmetric_pads.onnx")
```

Comment (alalek) [reply modules/dnn/test/test_onnx_importer.cpp:112]: Perhaps it makes sense to create new file like `generate_onnx_testdata_tf.py` (or tf2 to mention TensorFlow 2.x).

BTW, Old file can be renamed to `generate_onnx_testdata_pytorch.py` separately.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -1263,6 +1263,43 @@ void ONNXImporter::handleNode(const opencv_onnx::NodeProto& node_proto_)
...
+                    if (pads.get<int>(i) != pads.get<int>(i + dims))
+                    {
+                        asymmetricPadding = true;
+                        break;
+                    }
+                }
+                if (asymmetricPadding)
+                {
+                    layerParams.erase("pad");
+                    std::vector<int> paddings(pads.size());
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:1284]: Please add a comment that this line ends up not padding N and C out of NCHW.

File: modules/dnn/test/test_onnx_importer.cpp (hunk 1 above)
Comment (SamFC10) [reply modules/dnn/test/test_onnx_importer.cpp:112]: Sounds good. Should I add these 2 changes in opencv_extra PR?

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp:1284]: Isn't padding always done along H and W dimensions in convolution layer?

Comment (rogday) [reply modules/dnn/src/onnx/onnx_importer.cpp:1284]: I got confused looking at this line and thought there is an error and opened a debugger to understand why it's done this way. Maybe it's just me.

Comment (SamFC10) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Added comments in latest commit. Please check

Comment (rogday) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Looks good, thank you!

File: modules/dnn/test/test_onnx_importer.cpp (hunk 1 above)
Comment (alalek) [reply modules/dnn/test/test_onnx_importer.cpp:112]: Feel free to create a new file in opencv_extra.
It would be nice if you can prepare `requirements-tf.txt` file too (list required pip packages there, `pip freeze` may help with that, [example](https://github.com/opencv/opencv/blob/master/samples/dnn/dnn_model_runner/dnn_conversion/requirements.txt))

**Review by alalek — APPROVED** [review]
Looks good to me :+1:
Thank you for contribution!

**Comment by SamFC10** [comment]
Squashed commits into one
